File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/mutation/SemanticMutationEffect.java
package com.foundgine.core.semantic.mutation;

import com.foundgine.core.abstractions.*;

/** One semantic effect associated with a mutation operation. */
public record SemanticMutationEffect(SemanticMutationEffectKind kind, EntityId entity, FieldId field, RelationshipId relationship) {
    public SemanticMutationEffect { if (kind == null || entity == null) throw new NullPointerException(); }
    public SemanticMutationEffect(SemanticMutationEffectKind kind, EntityId entity) { this(kind, entity, null, null); }
    public SemanticMutationEffect(SemanticMutationEffectKind kind, EntityId entity, FieldId field) { this(kind, entity, field, null); }
}

File: src/csharp/Foundgine.Core/Semantic/Mutation/SemanticMutationEffect.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Mutation;

/// <summary>
/// One semantic effect associated with a mutation operation.
/// </summary>
public sealed record SemanticMutationEffect(
    SemanticMutationEffectKind Kind,
    EntityId Entity,
    FieldId? Field = null,
    RelationshipId? Relationship = null);
